cli/repl.rs

```diff
@@ -1,73 +1,240 @@
-    fs::create_dir_all(self.history_file.parent().unwrap())?;
-    self
-      .editor
-      .save_history(&self.history_file.to_str().unwrap())
-      .map(|_| debug!("Saved REPL history to: {:?}", self.history_file))
-      .map_err(|e| {
-        eprintln!("Unable to save REPL history: {:?} {}", self.history_file, e);
-        e.into()
-      })
-  }
+  let history_file = global_state.dir.root.join("deno_history.txt");
 
-  pub fn readline(&mut self, prompt: &str) -> Result<String, AnyError> {
-    self
-      .editor
-      .readline(&prompt)
-      .map(|line| {
-        self.editor.add_history_entry(line.clone());
-        line
-      })
-      .map_err(AnyError::from)
-
-    // Forward error to TS side for processing
-  }
-}
+  session
+    .post_message("Runtime.enable".to_string(), None)
+    .await?;
+
+  let helper = Helper {
+    validator: MatchingBracketValidator::new(),
+  };
+
+  let editor = Arc::new(Mutex::new(Editor::new()));
+
+  editor.lock().unwrap().set_helper(Some(helper));
+
+  editor
+    .lock()
+    .unwrap()
+    .load_history(history_file.to_str().unwrap())
+    .unwrap_or(());
+
+  println!("Deno {}", crate::version::DENO);
+  println!("exit using ctrl+d or close()");
+
+  let prelude = r#"
+    Object.defineProperty(globalThis, "_", {
+      configurable: true,
+      get: () => Deno[Deno.internal].lastEvalResult,
+      set: (value) => {
+       Object.defineProperty(globalThis, "_", {
+         value: value,
+         writable: true,
+         enumerable: true,
+         configurable: true,
+       });
+       console.log("Last evaluation result is no longer saved to _.");
+      },
+    });
+
+    Object.defineProperty(globalThis, "_error", {
+      configurable: true,
+      get: () => Deno[Deno.internal].lastThrownError,
+      set: (value) => {
+       Object.defineProperty(globalThis, "_error", {
+         value: value,
+         writable: true,
+         enumerable: true,
+         configurable: true,
+       });
+
+       console.log("Last thrown error is no longer saved to _error.");
+      },
+    });
+  "#;
 
-impl Drop for Repl {
-  fn drop(&mut self) {
-    self.save_history().unwrap();
+  session
+    .post_message(
+      "Runtime.evaluate".to_string(),
+      Some(json!({
+        "expression": prelude,
+        "contextId": context_id,
+      })),
+    )
+    .await?;
+
+  loop {
+    let editor2 = editor.clone();
+    let line = tokio::task::spawn_blocking(move || {
+      editor2.lock().unwrap().readline("> ")
+    })
+    .await?;
+
+    match line {
+      Ok(line) => {
+        let evaluate_response = session
+          .post_message(
+            "Runtime.evaluate".to_string(),
+            Some(json!({
+              "expression": format!("'use strict'; void 0;\n({})", line),
+              "contextId": context_id,
+              // TODO(caspervonb) set repl mode to true to enable const redeclarations and top
+              // level await
+              "replMode": false,
```

Comment:
It'd be great if we could switch this on before Friday's release

Reply:
I'll flip it and add tests as soon as this lands.